File: powerstrip/TestPowerStrip.py
import unittest
from PowerStrip import PowerStrip

class TestPowerStrip(unittest.TestCase):
    def _testInit(self,unit):
        self.assertEqual(unit.getMode(),PowerStrip.MODE_REMOTE)
        for outlet in range(unit.getOutlets()):
            self.assertEqual(unit.getOutletState(outlet),False,
                             "outlet " + str(outlet) + " should be off")
    def _testModes(self,unit):
        unit.setMode(PowerStrip.MODE_ON)
        for outlet in range(unit.getOutlets()):
            unit.setOutletState(outlet,False)
            self.assertEqual(unit.getOutletState(outlet),True)
            unit.setOutletState(outlet,True)
            self.assertEqual(unit.getOutletState(outlet),True)

        unit.setMode(PowerStrip.MODE_OFF)
        for outlet in range(unit.getOutlets()):
            unit.setOutletState(outlet,False)
            self.assertEqual(unit.getOutletState(outlet),False)
            unit.setOutletState(outlet,True)
            self.assertEqual(unit.getOutletState(outlet),False)

        unit.setMode(PowerStrip.MODE_REMOTE)
        for outlet in range(unit.getOutlets()):
            unit.setOutletState(outlet,False)
            self.assertEqual(unit.getOutletState(outlet),False)
            unit.setOutletState(outlet,True)
            self.assertEqual(unit.getOutletState(outlet),True)
        
        

    def _testModesInvalid(self,unit):
        for mode in ("remote","on","off","auto"):
            ok = False
            try:
                unit.setMode(mode)
            except ValueError:
                ok = True
            self.assertTrue(ok)

    def _testOutlet(self,unit,outlet):
        for state in (False,True):
            unit.setOutletState(outlet, state)
            unit.setMode(PowerStrip.MODE_REMOTE)
            self.assertEqual(unit.getOutletState(outlet), state,
                             "outlet " + str(outlet) + " should be " + str(state))
            unit.setMode(PowerStrip.MODE_OFF)
            self.assertEqual(unit.getOutletState(outlet),False,
                             "outlet " + str(outlet) + " in off mode should be False.")
            unit.setMode(PowerStrip.MODE_ON)
            self.assertEqual(unit.getOutletState(outlet),True,
                             "outlet " + str(outlet) + " in off mode should be True.")

    def _testOutletInvalid(self,unit,outlet):
        for state in (0.5,"off","on"):
            ok = False
            try:
                unit.setOutletState(outlet,state)
            except ValueError:
                ok = True
            self.assertTrue(ok, "outlet " + str(outlet) + " should fail to set value " + str(state))

    def _testOutlets(self,unit):
        for outlet in range(unit.getOutlets()):
            self._testOutlet(unit,outlet)
            self._testOutletInvalid(unit,outlet)
            
    def testConstructPowerStrip(self):
        unit = PowerStrip(2)
        self._testInit(unit)

    def testModesPowerStrip(self):
        unit = PowerStrip(2)
        self._testModes(unit)
        
    def testModesInvalidPowerStrip(self):
        unit = PowerStrip(2)
        self._testModesInvalid(unit)

    def testOutletsPowerStrip(self):
        unit = PowerStrip(2)
        self._testOutlets(unit)
        
if __name__ == '__main__':
    unittest.main()

File: powerstrip-cpp/src/testPowerStrip.cpp
#include <gtest/gtest.h>
#include <stdexcept>

#include "PowerStrip.h"

using namespace powerstrip;

TEST(PowerStrip, Construct) {
  for (int outlets = -1; outlets <= 4; ++outlets) {
    try {
      PowerStrip::Ptr ps(new PowerStrip(outlets));
      ASSERT_EQ(outlets,ps->getOutlets());
      ASSERT_EQ(Mode::REMOTE,ps->getMode());
      for (int i=-1; i<=ps->getOutlets(); ++i) {
        try {
          ASSERT_EQ(OutletState::UNPOWERED, ps->getOutletState(i));
        } catch (std::out_of_range &ex) {
          ASSERT_TRUE(i < 0 || i >= ps->getOutlets());

        }
      }
    } catch (std::out_of_range &ex) {
      ASSERT_TRUE(outlets <= 0);
    }
  }
}

TEST(PowerStrip, ModeOn) {
  int outlets = 4;
  powerstrip::PowerStrip ps(outlets);
  ps.setMode(powerstrip::Mode::ON);
  for (int outlet = 0; outlet < ps.getOutlets(); ++outlet) {
    ps.setOutletState(outlet,powerstrip::OutletState::POWERED);
    ASSERT_EQ(powerstrip::OutletState::POWERED,ps.getOutletState(outlet));
    ps.setOutletState(outlet,powerstrip::OutletState::UNPOWERED);
    ASSERT_EQ(powerstrip::OutletState::POWERED,ps.getOutletState(outlet));
  }
}

TEST(PowerStrip, ModeOff) {
  int outlets = 4;
  powerstrip::PowerStrip ps(outlets);
  ps.setMode(powerstrip::Mode::OFF);
  for (int outlet = 0; outlet < ps.getOutlets(); ++outlet) {
    ps.setOutletState(outlet,powerstrip::OutletState::POWERED);
    ASSERT_EQ(powerstrip::OutletState::UNPOWERED,ps.getOutletState(outlet));
    ps.setOutletState(outlet,powerstrip::OutletState::UNPOWERED);
    ASSERT_EQ(powerstrip::OutletState::UNPOWERED,ps.getOutletState(outlet));
  }
}

TEST(PowerStrip, ModeRemote) {
  int outlets = 4;
  powerstrip::PowerStrip ps(outlets);
  ps.setMode(powerstrip::Mode::REMOTE);
  for (int outlet = 0; outlet < ps.getOutlets(); ++outlet) {
    ps.setOutletState(outlet,powerstrip::OutletState::POWERED);
    ASSERT_EQ(powerstrip::OutletState::POWERED,ps.getOutletState(outlet));
    ps.setOutletState(outlet,powerstrip::OutletState::UNPOWERED);
    ASSERT_EQ(powerstrip::OutletState::UNPOWERED,ps.getOutletState(outlet));
  }
}





